opencl-aot/source/main.cpp

```diff
@@ -56,27 +57,41 @@ auto ContextDeleter = [](cl_context CLCon) {
   }
 };
 
-using CLContextUPtr = std::unique_ptr<
-    _cl_context, decltype(ContextDeleter)>; // decltype(CLContextUPtr.get()) is
-                                            // same as cl_context, should
-                                            // initialized with ContextDeleter
+using CLContextSPtr =
+    std::shared_ptr<_cl_context>; // decltype(CLContextSPtr.get()) is
+                                  // same as cl_context, should
+                                  // initialized with ContextDeleter
 
 /*! \brief Generate OpenCL program from OpenCL program binary (in ELF format) or
- * SPIR-V binary file
+ * SPIR-V binary file or LLVM IR (bitcode file) or OpenCL source file
  * \param FileNames (const std::vector<std::string>).
  * \param PlatformId (cl_platform_id).
  * \param DeviceId (cl_device_id).
+ * \param ContextSPtr (CLContextSPtr).
  * \return a tuple of vector of unique pointers to programs, error message and
  * return code.
  */
 std::tuple<std::vector<CLProgramUPtr>, std::string, cl_int>
-generateProgramsFromBinaries(const std::vector<std::string> &FileNames,
-                             cl_platform_id PlatformId, cl_device_id DeviceId) {
+generateProgramsFromInput(const std::vector<std::string> &FileNames,
+                          cl_platform_id PlatformId, cl_device_id DeviceId,
+                          CLContextSPtr ContextSPtr) {
   // step 0: define internal types
 
-  enum SupportedTypes : int8_t { BEGIN, ELF = BEGIN, SPIRV, UNKNOWN, END };
+  enum SupportedTypes : int8_t {
+    BEGIN,
+    ELF = BEGIN,
+    SPIRV,
+    LLVMIR,
```

Comment:
Can you add info about LLVM IR to help message?

Reply:
The LLVM IR objects are mostly like OpenCL compiled object files.

Reply:
Done.